Change password

Andrew-Ryer opened the pull request:
# Description

This PR implements the full User Password Change feature in OED, enabling any authenticated user to securely change their own password via the Options → Change Password menu, without admin intervention. The work spans backend API, frontend RTK Query integration, a new modal UI component, and integration into the main application with translations.

On the backend, it adds a new POST /api/users/changePassword endpoint protected by authentication middleware that verifies the user’s JWT, validates input (including length constraints), confirms the current password with bcrypt.compare, securely hashes the new password with bcrypt.hash, and updates the database via User.updateUserPassword. On the frontend, it introduces a changePassword RTK Query mutation and a ChangePasswordModal React component that displays the current username, collects current/new/confirm password, performs real-time validation, and shows success/error notifications. The feature is fully integrated into HeaderButtonsComponent via a “Change Password” entry in the Options dropdown and wired into the translations system so all UI text is available in English, French, and Spanish.

Fixes #1526 - Allow each user to change password

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the database configuration.
- [X] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [X] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [X] I have removed text in ( ) from the issue request
- [X] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations
- Password policy is limited to length-based validation (8–128 characters); there is no password strength meter or enforcement of complexity (e.g., symbols, digits, upper/lowercase mix).
- Password history / reuse prevention is not implemented; users can technically reuse previous passwords as long as they meet length requirements.
- No rate-limiting or lockout mechanism is added specifically for the password change endpoint; it relies on existing global protections.
- Automated tests (unit/integration/e2e) for the new endpoint, RTK Query mutation, and modal component are not included as part of this PR; verification has been described and performed manually and could be formalized in follow-up test work.
- Documentation and user-facing help text about the new “Change Password” option in the OED UI are not included and can be added in a separate documentation update if desired.

Contributors:
Andrew Ryer — @Andrew-Ryer — aryer@csumb.edu
Nathan Towsley — @natetowsley — natetowsley@gmail.com
Tyler Herndon — @GoKartMan89 — tylerherndon89@gmail.com
Caleb Stark — @caleb-stark — stark.caleba@gmail.com

natetowsley committed:
Added "Change Password" FormattedMessage to Options dropdown, added translations for "change.password" to data.ts

natetowsley committed:
Replaced previous modal state, for a new state, including ModalType to allow for the showing of more than one kind of modal. Adjusted handleClose() and handleShow() accordingly. Added a modalType to login modal and created a copy with modalType 'changePassword' as placeholder.

natetowsley committed:
Revert "Replaced previous modal state, for a new state, including ModalType to allow for the showing of more than one kind of modal. Adjusted handleClose() and handleShow() accordingly. Added a modalType to login modal and created a copy with modalType 'changePassword' as placeholder."

This reverts commit c20bec1da03ca9db215b5c3b90c26812588305af.

natetowsley committed:
,

natetowsley committed:
Change password modal component made, more translations added to data.ts, correct modal now shows when clicking change password.

natetowsley committed:
update hint message

natetowsley committed:
fix

caleb-stark committed:
fixed placeholder text

GoKartMan89 committed:
starting work on the userApi process

caleb-stark committed:
added translate

caleb-stark committed:
Merge branch 'nate_tests' of https://github.com/natetowsley/OED into nate_tests

Andrew-Ryer committed:
Backend API Development

Andrew-Ryer committed:
Merge branch 'nate_tests' of https://github.com/natetowsley/OED into nate_tests

natetowsley committed:
Implementation live share session.

Co-authored-by: GoKartMan89 <tylerherndon89@gmail.com>
Co-authored-by: caleb-stark <stark.caleba@gmail.com>
Co-authored-by: Andrew-Ryer <aryer@csumb.edu>

caleb-stark committed:
fixed terminal error for component props

GoKartMan89 committed:
Added all translations

GoKartMan89 committed:
Merge branch 'nate_tests' of https://github.com/natetowsley/OED into nate_tests

GoKartMan89 committed:
fixed a merge error

Andrew-Ryer committed:
Merge branch 'nate_tests' of https://github.com/natetowsley/OED into nate_tests

Andrew-Ryer committed:
Merge branch 'nate_tests' of https://github.com/natetowsley/OED into nate_tests

natetowsley committed:
removed middleware function from users.js

caleb-stark committed:
fixed the username formgroup for change password

caleb-stark committed:
added translation for username

caleb-stark committed:
removed unused import

natetowsley committed:
indentation/ format fix

natetowsley committed:
User now remains signed in when current password incorrectly entered into change password modal.

caleb-stark committed:
formatting/comments for change password files

caleb-stark committed:
resolved npm run check errors

huss commented on src/client/app/translations/data.ts:
```diff
@@ -633,6 +641,10 @@ const LocaleTranslationData = {
 		"calibration.submit.button": "Soumettre",
 		"cancel": "Annuler",
+		"change.password": "Changer le mot de passe",
+		"change.password.for": "Changer le mot de passe pour {username}",
```
I don't see this key in en nor does it seem used in the code. Is it needed?

huss commented on src/client/app/translations/data.ts:
```diff
@@ -444,6 +447,11 @@ const LocaleTranslationData = {
+		"password.new.enter": "Enter new password",
+		"password.successfully.changed": "Password changed successfully",
+		"password.failed.to.change": "Failed to change password",
+		"user.password.new": "New password",
```
I'm not seeing where this is used in the code. If kept, needs to start with password so correct alphabetically.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -48,6 +48,9 @@ const LocaleTranslationData = {
 		"calibration.save.database": "Save changes to database",
 		"calibration.submit.button": "Submit",
 		"cancel": "Cancel",
+		"change.password": "Change password",
```
I'm wondering if these should start with password as do a number of related ones such as ``password.new.enter``.

huss replied:
Thank you for changing ``change.password``. I'm thinking that ``current.password`` and ``current.password.enter`` might also be best if they start with password so in that section and consistent.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -1028,7 +1040,13 @@ const LocaleTranslationData = {
 		"page.restart.button": "Restart OED session\u{26A1}",
 		"page.user.refresh.directions": "If clicking the \"Return to Dashboard\" button does not work then please click the button below to restart your OED session\u{26A1}",
 		"password": "Mot de passe: ",
+		"password.change.username.label": "Nom d'utilisateur: {username}",
```
I don't see this in en nor used in the code. Is it needed?

huss commented on src/client/app/translations/data.ts:
```diff
@@ -444,6 +447,11 @@ const LocaleTranslationData = {
+		"password.confirm.enter": "Confirm new password",
+		"password.new.enter": "Enter new password",
+		"password.successfully.changed": "Password changed successfully",
+		"password.failed.to.change": "Failed to change password",
```
Off a few lines for alphabetical.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -1613,7 +1635,13 @@ const LocaleTranslationData = {
 		"page.user.refresh.directions": "If clicking the \"Return to Dashboard\" button does not work then please click the button below to restart your OED session\u{26A1}",
 		"password": "Contraseña: ",
+		"password.change.username.label": "Nombre de usuario: {username}",
+		"password.successfully.changed": "Contraseña cambiada exitosamente",
```
This and the next one are off alphabetically.

huss commented on src/server/routes/authenticator.js:
```diff
@@ -11,6 +11,7 @@ const validate = require('jsonschema').validate;
 const { isTokenAuthorized, isUserAuthorized } = require('../util/userRoles');
 const { getConnection } = require('../db');
 const escapeHtml = require('escape-html');
+const { resetWarningCache } = require('prop-types');
```
I'm not seeing where this import is used.

huss commented on src/server/routes/users.js:
```diff
@@ -3,13 +3,14 @@
 const validate = require('jsonschema').validate;
 const { getConnection } = require('../db');
 const jwt = require('jsonwebtoken');
+const { resetWarningCache } = require('prop-types');
```
I'm not seeing where this import is used.

huss commented on src/server/routes/authenticator.js:
```diff
@@ -205,12 +206,45 @@ optionalAuthMiddleware = (req, res, next) => {
+	} else if (token) {
+		jwt.verify(token, secretToken, async (err, decoded) => {
+			if (err) {
+				res.status(401).json({ success: false, message: 'Falied token verification'});
```
Falied -> Failed

huss commented on src/client/app/components/ChangePasswordModalComponent.tsx:
```diff
@@ -0,0 +1,227 @@
+				resetPasswordFields();
+			})
+			.catch(error => {
+				showErrorNotification(translate('password.failed.to.change') + ' ' + error.data.message);
```
I noticed that the login page says failed when the password is too short (< 8 characters) but here you get "invalid params" for the current password. I think they ought to be consistent. One option is to check it is at least the min required. Another is to use the way login does it of at least 1 character and the failed message. I think the former people chose the latter since it gives less info but I'm thinking a form warning might be nice since it is done for the new password.

huss commented on src/server/routes/users.js:
```diff
@@ -249,4 +250,56 @@ router.post('/delete', adminAuthMiddleware('delete a user'), async (req, res) =>
+		}
+	};
+if (!validate(req.body, validParams).valid) {
+	res.status(400).json({ message: 'Invalid params' });
```
I realize this is not done elsewhere in this file but the newer (and I think better) way is as in src/server/routes/conversions.js where it has:

```
	const validatorResult = validate(req.body, validConversion);
	if (!validatorResult.valid) {
		log.warn(`Got request to edit conversions with invalid conversion data, errors: ${validatorResult.errors}`);
		failure(res, 400, `Got request to edit conversions with invalid conversion data, errors: ${validatorResult.errors}`);
	} else {
```

This uses the standard way of returning the message and logs it which is a good idea for failed login attempts. What do you think?

natetowsley replied:
Everything worked with the newer method except no sub-message would show on client side underneath "Failed to change password" on the error notification when using failure(). I also tried to edit all routes to use this new method for consistency (still have the copy if that's worth looking into) but ultimately couldn't get it to work so replaced failure() line with res.status(400).

huss replied:
I was curious why it did not work so I tried to do it. I should soon push a commit to make the change. It also address some other items involving security that were outside the work you were asked to do. I welcome any thoughts you have about my changes.

huss commented on src/client/app/components/ChangePasswordModalComponent.tsx:
```diff
@@ -0,0 +1,227 @@
+	return (
+		<>
+			{/* Unsaved Warning Component */}
+			{showUnsavedWarning && (
```
I'm not seeing the warning even if I made changes. I have not looked into this.

huss replied:
I'm still seeing this issue. For example, on the edit user page for admins if you type any characters in either password field and then click off the modal it asks what to do in a popup. On this page it does not give the popup. Let me know if you don't see this or need any help.

caleb-stark replied:
This has been updated so now a warning will pop up when a user clicks off. This warning follows the same structure as the one on the admin edit user page but the rules for canSave were changed accordingly.

huss commented on src/client/app/components/HeaderButtonsComponent.tsx:
```diff
@@ -156,15 +157,17 @@ export default function HeaderButtonsComponent() {
+
 	// Handle modal visibility
-	const [showModal, setShowModal] = useState<boolean>(false);
+	type ModalType = 'login' | 'changePassword' | null;
```
It isn't necessary but I was thinking if a Typescript enum for the possible values makes sense. Then the logical name could be used throughout rather than repeating the string.

huss commented on src/server/routes/users.js:
```diff
@@ -249,4 +250,56 @@ router.post('/delete', adminAuthMiddleware('delete a user'), async (req, res) =>
+			}
+		}
+	};
+if (!validate(req.body, validParams).valid) {
```
I think the formatting/indenting is off here.

huss commented on src/server/routes/users.js:
```diff
@@ -249,4 +250,56 @@ router.post('/delete', adminAuthMiddleware('delete a user'), async (req, res) =>
+}
+try {
+	const conn = getConnection();
+	const userId = req.decoded.data; // From the authenticated token
```
Small point but OED prefers comments above the line of code.

huss reviewed (changes requested):
Thanks to @Andrew-Ryer, @natetowsley, @GoKartMan89 & @caleb-stark for this contribution to OED. I'm sorry it took about a week to get back to you. Overall, it works well. I have made some comments to consider. Please let me know if anything is not clear or I can help.

natetowsley committed:
Changed type ModalType into an enum for modal visibility.

GoKartMan89 committed:
changed HeaderButtonsComponent.tsx to update changed values in data.ts

GoKartMan89 committed:
Merge branch 'change_password' of https://github.com/natetowsley/OED into change_password

GoKartMan89 committed:
changed and cleaned up the added password-related
messages in the data.ts translation file

caleb-stark committed:
changed modal compoent error message inconsistancy

natetowsley committed:
Removal of unused import. Updated validation method used in /changePassword. Formatting and indentation fixes.

natetowsley committed:
"parameter" -> "parameters"

natetowsley committed:
Merge branch 'change_password' of https://github.com/natetowsley/OED into change_password

natetowsley committed:
Small indent fixes.

natetowsley committed:
Comment format fix.

natetowsley committed:
Typo addressed, unused import removed.

natetowsley commented:
@huss Changes addressed. Please let us know if there's anything else we should do.

huss committed:
Modify change password to use success/failure functions + other

- The messages are also displayed to the user.
- Changes information returned to mask info for security
reasons. Also logs more info for security.
- Minor changes in message wording for translation.
- Minor formatting changes.

huss reviewed (changes requested):
Thanks to @Andrew-Ryer, @natetowsley, @GoKartMan89 & @caleb-stark for the updated code to address my previous comments. There are a couple of old comments that I thought should still be addressed and I added to the comment chain. I also pushed a commit to use success/failure since you indicated issues in getting it to work. It also includes other changes to address ongoing security work that was not part of what you were doing. I would welcome you looking at the code to see if you think it is correct/appropriate and verifying it works as desired. Please let me know if anything is not clear or I can help.

huss commented:
For unknown reasons my recent comments are showing up twice. I'm just going to leave them to avoid any more issues. Sorry.

huss commented:
@Andrew-Ryer, @natetowsley, @GoKartMan89 & @caleb-stark It has been about a month since my last comments. I think this is close to complete so I wanted to know the current status. Thanks.

natetowsley commented:
@huss Apologies, I think all of us lost this is the midst of the school semester ending and the holidays coming around. I will have your changes addressed within the next week. Hope your holiday season went well.

natetowsley committed:
Merge most recent version of branch 'development' into change_password

natetowsley committed:
change.password and current.password.enter translations renamed for consistency.

caleb-stark committed:
added back the unsaved changes warning to change password

natetowsley commented:
Changes addressed @huss

huss commented on src/client/app/components/ChangePasswordModalComponent.tsx:
```diff
@@ -0,0 +1,247 @@
+							<Col>
+								<FormGroup>
+									<Label for='currentPassword'>
+										{translate('current.password')}
```
Should this now be password.current?

huss commented on src/client/app/components/ChangePasswordModalComponent.tsx:
```diff
@@ -0,0 +1,247 @@
+										id='currentPassword'
+										name='currentPassword'
+										type='password'
+										placeholder={translate('password.current.enter')}
```
I think this placeholder should be removed. The user creation page and other OED pages do no do this. The title is what OED uses to consistently indicate what to put in the box/item below it. The title is fine here. I also have comments in data.ts about the strings. The same applies to the other two inputs. I'm sorry I did not catch this earlier.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -446,7 +446,14 @@ const LocaleTranslationData = {
 		"password.confirm": "Confirm password: ",
+		"password.confirm.enter": "Confirm new password",
+		"password.current": "Current password",
+		"password.current.enter": "Enter current password",
```
This key can be removed in all languages. Note all changes to items should be done in all languages but I'm only noting that here.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -446,7 +446,14 @@ const LocaleTranslationData = {
+		"password.change": "Change password",
 		"password.confirm": "Confirm password: ",
+		"password.confirm.enter": "Confirm new password",
+		"password.current": "Current password",
```
Please put a colon (:) at the end so it is consistent with other titles.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -446,7 +446,14 @@ const LocaleTranslationData = {
 		"page.user.refresh.directions": "If clicking the \"Return to Dashboard\" button does not work then please click the button below to restart your OED session",
 		"password": "Password: ",
+		"password.change": "Change password",
 		"password.confirm": "Confirm password: ",
```
I think this would be better as "Confirm new password" (as the enter had but note there is no trailing space) so it is very clear and parallel to  "Current password". The next key can be removed.

caleb-stark replied:
"Change Password" is only used for the button in the menu I feel like it would be confusing to change it to "Confirm new Password" was this just a mistake you you meant to change "Confirm password" to "Confirm new password" i think it would involve a new key because confirm password is used in files separate from ours

huss replied:
Sorry for the delay in getting to this and that my comment was confusing. In another comment I suggested changing from "Password" to "New password". In the same way I was thinking to add "new" here so it is parallel on the web page. I also agree that it is only for that place on the web page so a new key may be needed. Does this help?

huss commented on src/client/app/translations/data.ts:
```diff
@@ -446,7 +446,14 @@ const LocaleTranslationData = {
+		"password.current": "Current password",
+		"password.current.enter": "Enter current password",
+		"password.failed.to.change": "Failed to change password with response: ",
+		"password.new.enter": "Enter new password",
```
This key can go.

huss commented on src/client/app/components/ChangePasswordModalComponent.tsx:
```diff
@@ -0,0 +1,247 @@
+							<Col>
+								<FormGroup>
+									<Label for='newPassword'>
+										{translate('password')}
```
I think this string should be "New password:" so it is very clear and consistent with other ones. This probably requires a new key in data.ts.

huss commented on src/client/app/components/ChangePasswordModalComponent.tsx:
```diff
@@ -0,0 +1,247 @@
+		const passwordModified = passwordDetails.currentPassword.length > 0 ||
+			passwordDetails.newPassword.length > 0 ||
+			passwordDetails.confirmPassword.length > 0;
+		
```
This file has a number of formatting issues. If you do 'format document' in VSC they should all go away. Please let me know if you need any help with this.

huss reviewed (changes requested):
Thanks to @Andrew-Ryer, @natetowsley, @GoKartMan89 & @caleb-stark for addressing the previous comments. I'm sorry it took a little while for me to look at this. It is working very well. I've made a few comments but I think they are all simple to do and then this should be ready to merge. Please let me know if you have thoughts or need anything.

huss commented:
@Andrew-Ryer, @natetowsley, @GoKartMan89 & @caleb-stark I'm trying to figure out the status of all pull requests and wondered if you have any input on this.

huss commented:
> @Andrew-Ryer, @natetowsley, @GoKartMan89 & @caleb-stark I'm trying to figure out the status of all pull requests and wondered if you have any input on this.

@Andrew-Ryer, @natetowsley, @GoKartMan89 & @caleb-stark This is a good PR that I would like to move forward. Can you give me any information on your interest/intent to address the comments? If not, I'll start the process of reassigning work on this PR to someone else on 6/13/26. I prefer to have you do it if you want.
